From: Miroslav Benes <mbenes@suse.cz>
Date: Mon, 30 Mar 2015 15:47:39 +0200
Subject: kgr: return error in kgr_init if notifier registration fails
Patch-mainline: submitted for review
References: fate#313296

Currently if registration of module notifier fails in kgr_init we just
warn and continue.  Notifier cleans up after the module when it leaves.
If it doesn't it can cause ftrace errors in the future. So we fail the
initialization if registration doesn't succeed.

Signed-off-by: Miroslav Benes <mbenes@suse.cz>
Reviewed-by: Petr Mladek <pmladek@suse.cz>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 kernel/kgraft.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

--- a/kernel/kgraft.c
+++ b/kernel/kgraft.c
@@ -1339,13 +1339,18 @@ static int __init kgr_init(void)
 	}
 
 	ret = register_module_notifier(&kgr_module_exit_nb);
-	if (ret)
-		pr_warn("Failed to register kGraft module exit notifier\n");
+	if (ret) {
+		pr_err("kgr: failed to register kGraft module exit notifier (%d)\n",
+			ret);
+		goto err_destroy_wq;
+	}
 
 	kgr_initialized = true;
 	pr_info("kgr: successfully initialized\n");
 
 	return 0;
+err_destroy_wq:
+	destroy_workqueue(kgr_wq);
 err_remove_files:
 	kgr_remove_files();
 
